ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.cache
+
+import com.daml.metrics.{Metrics, Timed}
+import com.daml.platform.store.cache.EventsBuffer.{
+  RequestOffBufferBounds,
+  UnorderedException,
+  SearchableByVector,
+}
+
+import scala.annotation.tailrec
+import scala.collection.Searching.{Found, InsertionPoint, SearchResult}
+import scala.math.Ordering
+import scala.math.Ordering.Implicits.infixOrderingOps
+
+/** An ordered-by-offset ring buffer.
+  *
+  * The buffer allows appending only elements with strictly increasing offsets, with the exception
+  * of a specially designated element, the range end marker,
+  * which can have an offset equal to the last appended element.
+  *
+  * @param maxBufferSize The maximum buffer size.
+  * @param metrics The DAML metrics.
+  * @param bufferQualifier The qualifier used for metrics tag specialization.
+  * @param isRangeEndMarker Identifies if an element [[E]] should be treated
+  *                         as a range end marker.
+  * @tparam O The offset type.
+  * @tparam E The entry buffer type.
+  */
+private[cache] final class EventsBuffer[O: Ordering, E](
+    maxBufferSize: Int,
+    metrics: Metrics,
+    bufferQualifier: String,
+    isRangeEndMarker: E => Boolean,
+) {
+  @volatile private var bufferRef = Vector.empty[(O, E)]
+
+  private var bufferRangeEnd: Option[O] = None
+
+  private val pushTimer = metrics.daml.services.index.buffer.push(bufferQualifier)
+  private val sliceTimer = metrics.daml.services.index.buffer.slice(bufferQualifier)
+  private val pruneTimer = metrics.daml.services.index.buffer.prune(bufferQualifier)
+
+  /** Appends a new event to the buffer.
+    *
+    * Starts evicting from the tail when `maxBufferSize` is reached.
+    *
+    * @param offset The event offset.
+    *              Must be higher than the last appended entry's offset, with the exception
+    *              of the range end marker, which can have an offset equal to the last appended element.
+    * @param entry The buffer entry.
+    */
+  def push(offset: O, entry: E): Unit =
+    Timed.value(
+      pushTimer,
+      synchronized {
+        if (bufferRef.size == maxBufferSize)
+          bufferRef = bufferRef.drop(1)
```

Comment:
This is probably a minor thing but in case when we have a full buffer and a range-end marker is pushed, a single buffer element will be dropped but no new element will be appended. Is that intentional?

Reply:
Good catch, thanks! Fortunately, this is harmless. I'll commit a fix